drivers/gp2y10xx/gp2y10xx.c

```diff
@@ -144,17 +104,21 @@ int gp2y10xx_read_density(gp2y10xx_t *dev, uint16_t *density)
     DEBUG("[gp2y10xx] read: aout=%u iled=%u\n", dev->params.aout,
           dev->params.iled_pin);
 
-    /* turn ILED on/off and wait a little bit to read the ADC */
-    _iled_on(dev);
-    xtimer_usleep(ILED_PULSE_WAIT_US);
-    if ((adc_value = adc_sample(dev->params.aout, dev->params.adc_res)) < 0) {
+    for (unsigned i = 0; i < CONFIG_GP2Y10XX_MAX_READINGS; i++) {
+        /* turn ILED on/off and wait a little bit to read the ADC */
+        _iled_on(dev);
+        xtimer_usleep(ILED_PULSE_WAIT_US);
```

Comment:
```suggestion
        xtimer_usleep(ILED_PULSE_WAIT_US);
        _iled_on(dev);
```

The way I understand the current ADC API, the samping should only start with `adc_sample()` - so you could turn off the LED while you wait for the next period.

Reply:
The sensor takes it's time in between the led gets powered on and detects particulate, so it's at the end of the 280 us that it generates data

Reply:
Then you can also move the `_iled_on(dev)` / `_iled_off(dev)` out of the loop

Reply:
Changin the iled_on/off out of the loop changes the output a little bit:

Current one:

```
2020-11-16 11:11:33,532 # Dust density 90 ug/m3
2020-11-16 11:11:33,771 # Dust density 85 ug/m3
2020-11-16 11:11:34,028 # Dust density 88 ug/m3
2020-11-16 11:11:34,283 # Dust density 83 ug/m3
2020-11-16 11:11:34,539 # Dust density 81 ug/m3
2020-11-16 11:11:34,796 # Dust density 90 ug/m3
```

Out of the loop:

```
2020-11-16 11:10:47,043 # Dust density 0 ug/m3
2020-11-16 11:10:47,299 # Dust density 424 ug/m3
2020-11-16 11:10:47,539 # Dust density 455 ug/m3
2020-11-16 11:10:47,795 # Dust density 424 ug/m3
2020-11-16 11:10:48,051 # Dust density 440 ug/m3
2020-11-16 11:10:48,291 # Dust density 448 ug/m3
2020-11-16 11:10:48,548 # Dust density 429 ug/m3
2020-11-16 11:10:48,803 # Dust density 435 ug/m3
2020-11-16 11:10:49,043 # Dust density 426 ug/m3
2020-11-16 11:10:49,299 # Dust density 451 ug/m3
```

Reply:
Huh that is weird - without debug output there should be almost no time between `_iled_off(dev)` and `_iled_on(dev)`

Reply:
In the [datasheet](https://www.sparkfun.com/datasheets/Sensors/gp2y1010au_e.pdf) they specify the 280 us sampling time, so IMO (while slower) should be left inside the loop.

![Screenshot from 2020-11-16 11-16-58](https://user-images.githubusercontent.com/5952531/99240810-4e5f2700-27fd-11eb-8af7-cd1966f860b3.png)

Reply:
Ah so it needs an on -> off transition to start sampling.
Then maybe add a short delay after `_iled_off(dev)` to make sure the off time is not too short on fast CPUs.

Reply:
Added a short 20us delay at the `_iled_off` just to be sure :eyes: :